src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -1005,6 +1006,97 @@ public void testExternalVersioningIndexConflictWithFlush() {
+        engine.delete(new Engine.Delete(index.type(), index.id(), index.uid()));
+        engine.forceMerge(true, 10, true, false); //expunge deletes
+
+        assertEquals(engine.segments(true).size(), 1);
+        try (Engine.Searcher test = engine.acquireSearcher("test")) {
+            assertEquals(numDocs-1, test.reader().numDocs());
+            assertEquals(numDocs-1, test.reader().maxDoc());
+        }
+
+        doc = testParsedDocument(Integer.toString(1), Integer.toString(1), "test", null, -1, -1, testDocument(), B_1, false);
+        index = new Engine.Index(null, newUid(Integer.toString(1)), doc);
+        engine.delete(new Engine.Delete(index.type(), index.id(), index.uid()));
+        engine.forceMerge(true, 10, false, false); //expunge deletes
+
+        assertEquals(engine.segments(true).size(), 1);
+        try (Engine.Searcher test = engine.acquireSearcher("test")) {
+            assertEquals(numDocs-2, test.reader().numDocs());
+            assertEquals(numDocs-1, test.reader().maxDoc());
+        }
+    }
+
+    public void testForceMergeAndClose() throws IOException, InterruptedException {
+        int numIters = randomIntBetween(2, 10);
+        for (int j = 0; j < numIters; j++) {
+            try (Store store = createStore()) {
+                final Translog translog = createTranslog();
+                final InternalEngine engine = createEngine(store, translog);
+                final CountDownLatch startGun = new CountDownLatch(1);
+                final CountDownLatch indexed = new CountDownLatch(1);
+
+                Thread thread = new Thread() {
+                    public void run() {
+                        try {
+                            try {
+                                startGun.await();
+                            } catch (InterruptedException e) {
+                                throw new RuntimeException(e);
+                            }
+                            int i = 0;
+                            while (true) {
+                                int numDocs  = randomIntBetween(1, 20);
+                                for (int j = 0; j < numDocs; j++) {
+                                    i++;
+                                    ParsedDocument doc = testParsedDocument(Integer.toString(i), Integer.toString(i), "test", null, -1, -1, testDocument(), B_1, false);
+                                    Engine.Index index = new Engine.Index(null, newUid(Integer.toString(i)), doc);
+                                    engine.index(index);
+                                }
+                                engine.refresh("test");
+                                indexed.countDown();
+                                try {
+                                    engine.forceMerge(randomBoolean(), 1, false, randomBoolean());
+                                } catch (OptimizeFailedEngineException ex) {
+                                    // ok
+                                    return;
```

Comment:
Did you confirm we sometimes hit this and not just ACE?  (The "indexed" CountDownLatch should make it likely...)